Recommendations (Jetpack Assistant): add unit tests to main React components

rcanepa opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to to-test.md in a new commit as part of your PR. -->

Related to 1200070516266605-as-1200070517042906

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
* Few small changes to the markup for better accessibility (read inline comments).
* Add unit tests to cover the behavior of the Recommendation module main components: `SiteType`, `FeaturePrompt`, and `Summary`.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

This doesn't bring changes to the product, it only adds unit tests to the Recommendations module.

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

No. It doesn't change what data or activity we track or use.

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

**Code review**
* Make sure code changes make sense to you and that tests actually cover the main logic of each tested component.

**Running tests**
* Go to the `projects/plugins/jetpack` directory (`cd projects/plugins/jetpack`).
* Execute `yarn test-gui`.
* Make sure all tests pass.

**Recommendation flow**
* Give a quick test to the whole Recommendation flow to make sure that the small changes made to markup didn't affect the original behavior of it.

rcanepa committed:
[not verified] Add missing key property

rcanepa committed:
[not verified] Add alt text to image for loading status

rcanepa committed:
[not verified] Spread props instead of passign the whole upsell object

rcanepa committed:
[not verified] Use constants rather than strings when working with steps

rcanepa committed:
[not verified] Add cover the Summary step with unit tests

rcanepa committed:
[not verified] Move fixtures to a separate file

rcanepa committed:
[not verified] Move test utils to its own file

rcanepa committed:
[not verified] Replace div with label for better accessibility

rcanepa committed:
[not verified] Improve PropTypes of the PromptLayout component

rcanepa committed:
[not verified] Improve test naming

rcanepa committed:
[not verified] Add unit tests to the SiteType component

rcanepa committed:
[not verified] Add unit tests to cover the continue button behavior

rcanepa committed:
[not verified] Move test utils to a better location

rcanepa committed:
[not verified] Make sure only checkbox's events are considered

rcanepa committed:
[not verified] Restore stubs after tests end

rcanepa committed:
[not verified] Add unit tests to the FeaturePrompt component

rcanepa committed:
[not verified] Use accessible markup to wrap sections

rcanepa committed:
[not verified] Add unit tests to the Content area

rcanepa commented on projects/plugins/jetpack/_inc/client/recommendations/prompts/checkbox-answer/index.jsx:
```diff
@@ -19,10 +19,16 @@ import { getDataByKey, updateRecommendationsData } from 'state/recommendations';
-	}, [ answerKey, checked, updateCheckboxAnswer ] );
+	const toggleCheckbox = useCallback(
+		e => {
+			if ( e.target.type !== 'checkbox' ) {
+				return;
+			}
```
Clicking on a `<label>` also triggers this event handler and we don't want it to run twice. We only want to run it when the target is of type `checkbox`.

rcanepa commented on projects/plugins/jetpack/_inc/client/recommendations/prompts/checkbox-answer/index.jsx:
```diff
@@ -44,9 +50,11 @@ const CheckboxAnswerComponent = ( { answerKey, checked, info, title, updateCheck
+				<input id={ answerKey } type="checkbox" defaultChecked={ checked } tabIndex={ -1 } />
 			</div>
-			<div className="jp-checkbox-answer__title">{ title }</div>
+			<label htmlFor={ answerKey } className="jp-checkbox-answer__title">
+				{ title }
+			</label>
```
Replaced `<div>` with `<label>` for better accessibility and testability.

rcanepa commented on projects/plugins/jetpack/_inc/client/recommendations/summary/index.jsx:
```diff
@@ -72,30 +72,34 @@ const SummaryComponent = props => {
-						</p>
-					) }
-				</div>
+				<section aria-labelledby="enabled-recommendations">
+					<h2 id="enabled-recommendations">{ __( 'Recommendations enabled', 'jetpack' ) }</h2>
+					<div>
+						{ summaryFeatureSlugs.selected.length > 0 ? (
+							summaryFeatureSlugs.selected.map( slug => (
+								<FeatureSummary key={ slug } featureSlug={ slug } />
+							) )
+						) : (
+							<p>
+								<em>
+									{ __(
+										'You didn’t enable any recommended features. To get the most out of Jetpack, enable some recommendations or explore all Jetpack features.',
+										'jetpack'
+									) }
+								</em>
+							</p>
+						) }
+					</div>
+				</section>
```
Wrapping the enabled features in a section for better accessibility.

rcanepa commented on projects/plugins/jetpack/_inc/client/recommendations/summary/index.jsx:
```diff
@@ -72,30 +72,34 @@ const SummaryComponent = props => {
-						<h2>{ __( 'Recommendations skipped', 'jetpack' ) }</h2>
+					<section aria-labelledby="skipped-recommendations">
+						<h2 id="skipped-recommendations">{ __( 'Recommendations skipped', 'jetpack' ) }</h2>
 						<div>
 							{ summaryFeatureSlugs.skipped.map( slug => (
-								<FeatureSummary featureSlug={ slug } />
+								<FeatureSummary key={ slug } featureSlug={ slug } />
 							) ) }
 						</div>
-					</>
+					</section>
```
Wrapping the skipped features in a section for better accessibility.

rcanepa committed:
[not verified] Remove unnecessary comment

rcanepa commented on projects/plugins/jetpack/_inc/client/test/test-utils.js:
```diff
@@ -0,0 +1,29 @@
+export * from '@testing-library/react';
+
+// override render method
+export { render };
```
This is a pretty standard utils file that is used to create a render function that provides a Redux store to the component being tested. In short, it makes testing connected components much easier.

rcanepa committed:
[not verified] Stub the updateRecommendationsStep action on each test file

rcanepa committed:
[not verified] Add missing piece of state

rcanepa committed:
[not verified] Add changelog file

rcanepa commented:
I don't know how to fix the `All commits were linted before commit.` warning. It seems that when I rebase the PR, all pre-commit checks are skipped.

anomiex commented:
There isn't really a good way to do it. As long as the various "Linting" CI checks are passing it should be ok to ignore it; the main thing potentially missing is the "prettier" reformatting of JS and JSON files, which might result in a future PR editing those files having extraneous formatting changes.

The best way I've come up with to clear the warning is to do an interactive rebase (`git rebase -i origin/master`), change every commit from "pick" to "reword", and edit the commit messages to remove the `[not verified]`.

rcanepa commented:
Rebased to solve merge conflicts.

adimoldovan commented on projects/plugins/jetpack/_inc/client/recommendations/prompts/feature-prompt/test/component.js:
```diff
@@ -0,0 +1,171 @@
+		} );
+
+		it( 'calls the right actions when a user clicks on the skip feature button', () => {
+			const DUMMY_ACTION = { type: 'dummy' };
```
This is already declared on line 42

```suggestion

```

rcanepa replied:
Good catch! I'm definitely removing that unneccesary `const` definition.

adimoldovan reviewed (dismissed):
Looks good to me! I only left a minor suggestion.

rcanepa committed:
[not verified] Remove duplicated const definition

rcanepa commented:
Hey @adimoldovan, I removed the duplicated `const` definition (and rebased the PR) but Github dismissed your approval. Can I have another one after all checks are completed?

rcanepa committed:
Merge branch 'master' into add/unit-tests-to-recommendation-module
